msrest/polling/poller.py

```diff
@@ -0,0 +1,214 @@
+        pass
+
+    def status(self):
+        """Return the current status as a string.
+        :rtype: str
+        """
+        return "succeeded"
+
+    def finished(self):
+        """Is this polling finished?
+        :rtype: bool
+        """
+        return True
+
+    def resource(self):
+        return self._deserialization_callback(self._initial_response)
+
+
+class LROPoller(object):
+    """Poller for long running operations.
+
+    :param client: A msrest service client. Can be a SDK client and it will be casted to a ServiceClient.
+    :type client: msrest.service_client.ServiceClient
+    :param initial_response: The initial call response
+    :type initial_response: requests.Response or msrest.pipeline.ClientRawResponse
+    :param deserialization_callback: A callback that takes a Response and return a deserialized object. If a subclass of Model is given, this passes "deserialize" as callback.
+    :type deserialization_callback: callable or msrest.serialization.Model
+    :param polling_method: The polling strategy to adopt
+    :type polling_method: msrest.polling.PollingMethod
+    """
+
+    def __init__(self, client, initial_response, deserialization_callback, polling_method):
+        try:
+            self._client = client if isinstance(client, ServiceClient) else client._client
+        except AttributeError:
+            raise ValueError("Poller client parameter must be a low-level msrest Service Client or a SDK client.")
+        self._response = initial_response.response if isinstance(initial_response, ClientRawResponse) else initial_response
+        self._callbacks = []
+        # polling_method attribute is public (to be accessed from the callback)
+        self._polling_method = polling_method
+
+        if isinstance(deserialization_callback, type) and issubclass(deserialization_callback, Model):
+            deserialization_callback = deserialization_callback.deserialize
+
+        # Might raise a CloudError
+        self._polling_method.initialize(self._client, self._response, deserialization_callback)
+
+        # Prepare thread execution
+        self._thread = None
+        self._done = None
+        self._exception = None
+        if not self._polling_method.finished():
+            self._done = threading.Event()
+            self._thread = threading.Thread(
+                target=self._start,
+                name="LROPoller({})".format(uuid.uuid4()))
+            self._thread.daemon = True
+            self._thread.start()
+
+    def _start(self):
+        """Start the long running operation.
+        On completion, runs any callbacks.
+
+        :param callable update_cmd: The API reuqest to check the status of
+         the operation.
+        """
+        try:
+            self._polling_method.run()
+        except Exception as err:
+            self._exception = err
+
+        finally:
+            self._done.set()
+
+        callbacks, self._callbacks = self._callbacks, []
+        while callbacks:
```

Comment:
Is this loop necessary given that we do not allow callbacks to be added or removed once self._done has been set?

Reply:
I'm not sure why this "while" was here in the first place. What kind of pattern did you implement?

Reply:
I guess the code is trying to secure concurrent access to `self.__callbacks` and put them back after callback execution.